core/trino-main/src/main/java/io/trino/operator/function/StreamingTableFunctionInputProvider.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.operator.function;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.primitives.Ints;
+import io.trino.operator.WorkProcessor;
+import io.trino.spi.Page;
+
+import java.util.List;
+import java.util.Optional;
+
+public class StreamingTableFunctionInputProvider
+        implements WorkProcessor.Transformation<Page, List<Optional<Page>>>
+{
+    private boolean processEmptyInput;
+    private final int[][] requiredChannels;
+
+    public StreamingTableFunctionInputProvider(boolean processEmptyInput, List<List<Integer>> requiredChannels)
+    {
+        this.processEmptyInput = processEmptyInput;
+        this.requiredChannels = requiredChannels.stream()
+                .map(Ints::toArray)
+                .toArray(int[][]::new);
+    }
+
+    @Override
+    public WorkProcessor.TransformationState<List<Optional<Page>>> process(Page input)
+    {
+        if (input == null) {
+            if (processEmptyInput) {
+                // it can only happen at the first call to process(), which implies that there is no input. Empty PagesIndex can be passed on only once.
+                processEmptyInput = false;
+                return WorkProcessor.TransformationState.ofResult(null, false);
+            }
+            return WorkProcessor.TransformationState.finished();
+        }
+
+        // there is input, so we are not interested in processing empty input
+        processEmptyInput = false;
+
+        ImmutableList.Builder<Optional<Page>> sourcePages = ImmutableList.builder();
+
+        for (int[] requiredChannel : requiredChannels) {
+            sourcePages.add(Optional.of(input.getColumns(requiredChannel)));
```

Comment:
When there are multiple sources, the logic of splitting input into sources is more complicated than this. See `RegularTableFunctionPartition.prepareInputPages()`.
For now, just assert there is only one source.